pkg/middleware/auth_proxy_test.go

```diff
@@ -0,0 +1,85 @@
+package middleware
+
+import (
+	"fmt"
+	"net/http"
+	"testing"
+
+	"github.com/grafana/grafana/pkg/bus"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/infra/remotecache"
+	authproxy "github.com/grafana/grafana/pkg/middleware/auth_proxy"
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/setting"
+	"github.com/stretchr/testify/require"
+	macaron "gopkg.in/macaron.v1"
+)
+
+// Test initContextWithAuthProxy with a cached user ID that is no longer valid.
+//
+// In this case, the cache entry should be ignored/cleared and another attempt should be done to sign the user
+// in without cache.
+func TestInitContextWithAuthProxy_CachedInvalidUserID(t *testing.T) {
+	const name = "markelog"
+	const userID = int64(1)
+	const orgID = int64(4)
+
+	upsertHandler := func(cmd *models.UpsertUserCommand) error {
+		require.Equal(t, name, cmd.ExternalUser.Login)
+		cmd.Result = &models.User{Id: userID}
+		return nil
+	}
+	getSignedUserHandler := func(cmd *models.GetSignedInUserQuery) error {
+		// Simulate that the cached user ID is stale
+		if cmd.UserId != userID {
+			return models.ErrUserNotFound
+		}
+
+		cmd.Result = &models.SignedInUser{
+			UserId: userID,
+			OrgId:  orgID,
+		}
+		return nil
+	}
+
+	origHeaderName := setting.AuthProxyHeaderName
+	origEnabled := setting.AuthProxyEnabled
+	origHeaderProperty := setting.AuthProxyHeaderProperty
+	t.Cleanup(func() {
+		setting.AuthProxyHeaderName = origHeaderName
+		setting.AuthProxyEnabled = origEnabled
+		setting.AuthProxyHeaderProperty = origHeaderProperty
+		bus.RemoveHandler(upsertHandler)
+		bus.RemoveHandler(getSignedUserHandler)
```

Comment:
@papagian I think it's better to not make assumptions. It's a rule of thumb that tests shouldn't affect the state of other tests.

Reply:
ok, but in that case instead of removing completely the handler I think we need to restore its previous behaviour; I believe that tests expecting to find that handler would fail [here](https://github.com/grafana/grafana/blob/53175a41c9e6becc0b493d3cadc8ade2524aba27/pkg/bus/bus.go#L119)

Reply:
@papagian I found a way of restoring any previous handlers (they're currently `nil` AFAICT). Please have another look.